crates/pathfinder/src/retry.rs

```diff
@@ -0,0 +1,351 @@
+//! A general utility for retrying futures with a configurable backoff and error filter.
+use std::{
+    future::Future,
+    num::{NonZeroU64, NonZeroUsize},
+    result::Result,
+    time::Duration,
+};
+use tokio_retry::{strategy::ExponentialBackoff, Retry as TokioRetry, RetryIf as TokioRetryIf};
+
+pub struct Retry<T, E, Fut, FutureFactory>
+where
+    Fut: Future<Output = Result<T, E>>,
+    FutureFactory: FnMut() -> Fut,
+{
+    future_factory: FutureFactory,
+    strategy: Strategy,
+}
+
+impl<T, E, Fut, FutureFactory> Retry<T, E, Fut, FutureFactory>
+where
+    Fut: Future<Output = Result<T, E>>,
+    FutureFactory: FnMut() -> Fut,
+{
+    /// Create an exponential [`Retry`] utility for a future which is created by `future_factory`
+    /// with initial backoff of `base_secs` seconds.
+    ///
+    /// `Nth` backoff is equal to `base_secs ^ N` seconds.
+    pub fn exponential(future_factory: FutureFactory, base_secs: NonZeroU64) -> Self {
+        Self {
+            future_factory,
+            strategy: Strategy {
+                initial_backoff_secs: base_secs,
+                factor: NonZeroU64::new(1).unwrap(),
+                max_delay_secs: None,
+                max_num_retries: None,
+            },
+        }
+    }
+
+    /// Multiply backoff by this factor.
+    ///
+    /// `Nth` backoff is then equal to `base_secs ^ N * factor` seconds.
+    pub fn factor(mut self, factor: NonZeroU64) -> Self {
+        self.strategy.factor = factor;
+        self
+    }
```

Comment:
I'm a bit worried about this formula, I had similar concerns when looking at `tokio-retry`.

Do we not want some exp factor like `1.5`? Having squaring being the lowest form seems.. rough.

Reply:
The equation is `(base ^ N) * factor`, where base is in seconds. I would suggest that we then allow for a floating point base which is > 1.0.